test-clients/src/main/java/com/hedera/services/bdd/spec/util/EncodingUtil.java

```diff
@@ -0,0 +1,125 @@
+                convertedArgumentTypes[i] = "int64";
+            } else if(params[i] instanceof BigInteger && !("uint64".equals(splittedArgumentTypes[i])) || "int128".equals(splittedArgumentTypes[i])
+                    || "uint128".equals(splittedArgumentTypes[i]) || "int256".equals(splittedArgumentTypes[i]) || "uint256".equals(splittedArgumentTypes[i])) {
+                convertedArgumentTypes[i] = "uint256";
+            } else {
+                convertedArgumentTypes[i] = splittedArgumentTypes[i];
+            }
+        }
+
+        return "("  + Arrays.stream(convertedArgumentTypes).collect(Collectors.joining(",")).concat(")");
+    }
+
+    private static Object[] convertAddressToHeadlong(final String abi, final Object[] params) {
+        Object[] result = new Object[params.length];
+        for (int i = 0; i < params.length; i++) {
+//            if(params[i] instanceof String && ((String) params[i]).length() > 10){
+//                result[i] = convertAliasToHeadlongAddress((String) params[i]);
+//            } else
+            if(params[i] instanceof byte[] && (((byte[]) params[i]).length == 20)) {
+                result[i] = asHeadlongAddress((byte[]) params[i]);
+            } else if(params[i] instanceof byte[] && (((byte[]) params[i]).length == 40)) {
+                result[i] = convertAliasToHeadlongAddress((String) params[i]);
+            } else if (params[i] instanceof List) {
+                if(!((List<?>) params[i]).isEmpty() && ((List<?>) params[i]).get(0) instanceof byte[]){
+                    Address[] objectsInside = new Address[((List<?>) params[i]).size()];
+                    for (int j = 0; j < ((List<?>) params[i]).size(); j++) {
+                        if(((byte[]) ((List<?>) params[i]).get(j)).length == 20) {
+                            var instance = ((List<?>) params[i]).get(j);
+                            instance = asHeadlongAddress((byte[]) instance);
+                            objectsInside[j] = (Address) instance;
+                        }
+                        if(((byte[]) ((List<?>) params[i]).get(j)).length == 40) {
+                            var instance = ((List<?>) params[i]).get(j);
+                            instance = convertAliasToHeadlongAddress((String) instance);
+                            objectsInside[j] = (Address) instance;
+                        }
+                    }
+                    result[i] = objectsInside;
+                } else {
+//                    //TODO: Find a way to crate arrays from List
+//                    ((List<?>) params[i]).toArray();
+//                    for (int j = 0; j < ((List<?>) params[i]).size(); j++) {
+//
+//                    }
+                    result[i] = ((List<?>) params[i]).stream().mapToLong(l -> (long) l).toArray();
+                }
+            } else {
+                result[i] = params[i];
+            }
+        }
+
+        return result;
+    }
```

Comment:
This is pretty complicated and unreadable... it's also prone to errors. 
We should research cleaner ways of encoding and decoding parameters and try to make it more straightforward to read and write.
The headlong library itself should be responsible for handling the complexity and abstracting it from our code. 
We know what function we're calling at the point when we're trying to encode the params. We should be able to do something like this:
```
Function f = new Function("baz(uint32,bool)"); // canonicalizes and parses any signature
ByteBuffer one = f.encodeCall(args);
```
This is copied from the `headlong` [docs](https://github.com/esaulpaugh/headlong).
It should be that simple and not require any of this complicated logic. If we need to, we can have a number of util methods for passing `alias` or `mirror` addresses as args, but they shouldn't be that complicated either.